python/examples/contraction/definitions.py

```diff
@@ -45,33 +49,37 @@ def keys(self) -> List[str]:
 
   def shapes_builder(self, sizes: Mapping[str, Any]) -> List[List[int]]:
     """Constructs the tensor shapes given problem parameters."""
+    operand_dims = [
+      self.specification.lhs_dims,
+      self.specification.rhs_dims,
+      self.specification.output_dims]
 
-    def shape_of_tensor(name: str):
-      return [sizes[d] for d in getattr(self.specification, name)]
+    def shape_of_tensor(dims: str):
+      return [sizes[k] for k in dims]
 
-    return [
-        shape_of_tensor(name)
-        for name in ['lhs_dims', 'rhs_dims', 'output_dims']
-    ]
+    return [shape_of_tensor(d) for d in operand_dims if d is not None]
 
   def gflop_count_builder(self, sizes: Mapping[str, Any]) -> float:
     """Returns the GFLOp count given problem parameters."""
-    return 2.0 * np.prod(list(sizes.values())) / 1.e9
+    return self.flop_count * np.prod(list(sizes.values())) / 1.e9
 
   def gbyte_count_builder(self, sizes: Mapping[str, Any],
                           types: Sequence[np.dtype]) -> float:
     """Return the GByte count given problem parameters."""
-    lhs_type, rhs_type, output_type = types
-    lhs_shape, rhs_shape, output_shape = self.shapes_builder(sizes)
-    ro_gbytes = 1.e-9 * (np.prod(lhs_shape) * np.dtype(lhs_type).itemsize +
-                         np.prod(rhs_shape) * np.dtype(rhs_type).itemsize)
-    rw_gbytes = 2.e-9 * np.prod(output_shape) * np.dtype(output_type).itemsize
-    return ro_gbytes + rw_gbytes
+    shapes = self.shapes_builder(sizes)
+    gbyte_count = 0
+    for type, shape in zip(types, shapes):
+      gbyte_count += 1.e-9 * (np.prod(shape) * np.dtype(type).itemsize)
+    # Reduction operators read and write the output.
+    if self.specification.reduction_dims:
+      gbyte_count += 1.e-9 * \
+          (np.prod(shapes[-1]) * np.dtype(types[-1]).itemsize)
+    return gbyte_count
 
   def tensors_np_builder(self, sizes: Mapping[str, Any],
                          types: Sequence[np.dtype]) -> List[np.dtype]:
     """Returns random NumPy suitable for calling the kernel."""
-    shapes = self.shapes_builder(sizes)
+    shapes = [s if s else [1] for s in self.shapes_builder(sizes)]
```

Comment:
I do not follow this change, why is `[1]` the right answer? Is it because it is just ignored ? (maybe add some extra doc for the future?)

Reply:
Ah this is to handle the scalar case. If we do not have a size we want to get a numpy scalar tensor which has size one. It is not new code we use it in other parts of the code base.

Reply:
ok